File: Sources/SimConsole/Mock.swift
import Foundation

/// One mock rule. Persisted as a row in `~/.sim-console/mocks-<bundle-id>.json`.
public struct Mock: Codable, Identifiable, Equatable {
    public var id: String
    public var match: MockMatch
    public var response: MockResponse
    public var delayMs: Int
    public var enabled: Bool
    public var createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, match, response, enabled
        case delayMs = "delay_ms"
        case createdAt = "created_at"
    }

    public init(
        id: String = UUID().uuidString,
        match: MockMatch,
        response: MockResponse,
        delayMs: Int = 0,
        enabled: Bool = true,
        createdAt: String = ISO8601DateFormatter().string(from: Date())
    ) {
        self.id = id
        self.match = match
        self.response = response
        self.delayMs = delayMs
        self.enabled = enabled
        self.createdAt = createdAt
    }
}

public struct MockMatch: Codable, Equatable {
    public var method: String
    public var url: String
    /// Forward-compatible: when set, the request body must contain this
    /// substring for the mock to apply. v1 UI does not expose this; the
    /// iOS side honors it so power-users editing the JSON by hand can use it.
    public var bodyContains: String?

    enum CodingKeys: String, CodingKey {
        case method, url
        case bodyContains = "body_contains"
    }

    public init(method: String, url: String, bodyContains: String? = nil) {
        self.method = method
        self.url = url
        self.bodyContains = bodyContains
    }

    /// Exact match on method + full URL. Optionally requires the request body
    /// to contain `bodyContains`.
    public func matches(_ request: URLRequest) -> Bool {
        guard let url = request.url?.absoluteString else { return false }
        guard method.uppercased() == (request.httpMethod ?? "GET").uppercased() else { return false }
        guard url == self.url else { return false }
        if let needle = bodyContains, !needle.isEmpty {
            let body = MockMatch.requestBodyString(request) ?? ""
            return body.contains(needle)
        }
        return true
    }

    fileprivate static func requestBodyString(_ request: URLRequest) -> String? {
        if let data = request.httpBody {
            return String(data: data, encoding: .utf8)
        }
        if let stream = request.httpBodyStream {
            stream.open()
            defer { stream.close() }
            var data = Data()
            var buffer = [UInt8](repeating: 0, count: 4096)
            while stream.hasBytesAvailable {
                let n = stream.read(&buffer, maxLength: buffer.count)
                if n <= 0 { break }
                data.append(buffer, count: n)
            }
            return String(data: data, encoding: .utf8)
        }
        return nil
    }
}

public struct MockResponse: Codable, Equatable {
    public var status: Int
    public var headers: [String: String]
    public var body: String?

    public init(status: Int, headers: [String: String] = [:], body: String? = nil) {
        self.status = status
        self.headers = headers
        self.body = body
    }
}

/// File-level wrapper persisted as JSON.
struct MockFile: Codable {
    var version: Int
    var mocks: [Mock]
}

File: android/simconsole/src/main/kotlin/com/simconsole/Mock.kt
package com.simconsole

import org.json.JSONArray
import org.json.JSONObject

/**
 * One mock rule. Persisted as a row in the panel-managed JSON file (see
 * [MockStore]).
 *
 * Field names and types must stay byte-for-byte compatible with the iOS Mock
 * struct in `Sources/SimConsole/Mock.swift`: the macOS panel writes one file
 * format and both SDKs read it. The iOS UI exposes `body_contains` only via
 * manual edits, but both SDKs honor it.
 */
data class Mock(
    val id: String,
    val match: MockMatch,
    val response: MockResponse,
    val delayMs: Int = 0,
    val enabled: Boolean = true,
    val createdAt: String = "",
) {
    internal companion object {
        fun fromJson(obj: JSONObject): Mock = Mock(
            id = obj.getString("id"),
            match = MockMatch.fromJson(obj.getJSONObject("match")),
            response = MockResponse.fromJson(obj.getJSONObject("response")),
            delayMs = obj.optInt("delay_ms", 0),
            enabled = obj.optBoolean("enabled", true),
            createdAt = obj.optString("created_at", ""),
        )
    }
}

data class MockMatch(
    val method: String,
    val url: String,
    val bodyContains: String? = null,
) {
    /** Exact match on method + full URL, optionally requiring the body substring. */
    fun matches(method: String, url: String, body: String?): Boolean {
        if (!this.method.equals(method, ignoreCase = true)) return false
        if (this.url != url) return false
        val needle = bodyContains
        if (!needle.isNullOrEmpty()) {
            return body?.contains(needle) == true
        }
        return true
    }

    internal companion object {
        fun fromJson(obj: JSONObject): MockMatch = MockMatch(
            method = obj.getString("method"),
            url = obj.getString("url"),
            bodyContains = obj.optString("body_contains").ifEmpty { null },
        )
    }
}

data class MockResponse(
    val status: Int,
    val headers: Map<String, String> = emptyMap(),
    val body: String? = null,
) {
    internal companion object {
        fun fromJson(obj: JSONObject): MockResponse {
            val headers = mutableMapOf<String, String>()
            obj.optJSONObject("headers")?.let { hdr ->
                for (key in hdr.keys()) headers[key] = hdr.getString(key)
            }
            val bodyStr = if (obj.has("body") && !obj.isNull("body")) obj.optString("body") else null
            return MockResponse(
                status = obj.getInt("status"),
                headers = headers,
                body = bodyStr,
            )
        }
    }
}

/** Top-level JSON file shape. Mirrors iOS `MockFile`. */
internal data class MockFile(val version: Int, val mocks: List<Mock>) {
    companion object {
        fun parse(text: String): MockFile {
            val obj = JSONObject(text)
            val arr = obj.optJSONArray("mocks") ?: JSONArray()
            val mocks = (0 until arr.length()).map { Mock.fromJson(arr.getJSONObject(it)) }
            return MockFile(version = obj.optInt("version", 1), mocks = mocks)
        }
    }
}
